### PR TITLE
Add <Common.MarkdownEditor /> component

### PR DESCRIPTION
**Pull request by kntsoriano**
## Description
This adds the markdown editor component, which will be reused throughout the site.

## Screenshot
Screen Shot 2020-12-18 at 11 59 16 AM

## Questions
1. What options do we use from [here](https://github.com/Ionaru/easy-markdown-editor#options-list)?
2. Do we have a final design for the preview tab? For now, I just used the preview button (eye icon) in the markdown editor.

### CODE REVIEW

File: app/javascript/components/common/MarkdownEditor.tsx [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+          if (!url) {
+            return 'Preview unavailable'
+          }
+
+          sendRequest({
+            endpoint: url,
+            body: JSON.stringify({ markdown: markdown }),
+            method: 'POST',
+            isMountedRef: isMountedRef,
+          }).then((json: any) => {
```
Comment (ErikSchierboom) [inline app/javascript/components/common/MarkdownEditor.tsx]: What happens when there is an error returned by the HTTP request? Will it keep on showing 'Loading...'?

File: app/javascript/components/common/MarkdownEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
+import React from 'react'
+import SimpleMDE from 'react-simplemde-editor'
+import 'easymde/dist/easymde.min.css'
+import { sendRequest } from '../../utils/send-request'
+import { useIsMounted } from 'use-is-mounted'
+
+export const MarkdownEditor = ({ uuid }: { uuid: string }): JSX.Element => {
```
Comment (ErikSchierboom) [inline app/javascript/components/common/MarkdownEditor.tsx]: I was initially wondering why the markdown editor needs a uuid :D

**Review by ErikSchierboom — APPROVED** [review]
Do we need some integration tests to verify that markdown in the component is actually rendered as HTML?

File: app/javascript/components/common/MarkdownEditor.tsx (hunk 2 above)
Comment (iHiD) [reply app/javascript/components/common/MarkdownEditor.tsx:7]: Why does it? :)

File: app/views/test/components/common/markdown_editor/show.html.haml [hunk 3]
```diff
@@ -0,0 +1 @@
+= ReactComponents::Common::MarkdownEditor.new(SecureRandom.uuid)
```
Comment (iHiD) [inline app/views/test/components/common/markdown_editor/show.html.haml]: What's the purpose of the `uuid` here? If it's just random each time, then this should be moved into the `Component` class. 

```suggestion
= ReactComponents::Common::MarkdownEditor.new(SecureRandom.uuid)
```

**Comment by iHiD** [comment]
Re the ID: I see now that it's used for the auto-save. I think my confusion is that it's called `uuid` (which is a type) rather than `id` (which is a property). So I think we just need a more descriptive name. Maybe `context_id` would make the most sense, as we're scoping it to some context (e.g. We might use `"mentor_discussion:#{discussion.uuid}"` as the id on a mentoring discussion page, or `bug_report_new` on that page.

---

1. What options do we use from here?

- `autosave:` enabled (You have all this set up)
- `blockStyles/italic:` `_`

I think beyond those we should just see how it works then improve it form there.

Nice work :)

File: app/controllers/api/markdown_controller.rb [hunk 4]
```diff
@@ -0,0 +1,9 @@
+module API
+  class MarkdownController < BaseController
+    def parse
+      html = ParseMarkdown.(params[:markdown].to_s)
+
+      render json: { html: html }
+    end
+  end
+end
```
Comment (iHiD) [inline app/controllers/api/markdown_controller.rb:9]: Pls add a controller test that checks calling it with markdown returns html and is a 200.

File: app/views/layouts/application.html.haml [hunk 5]
```diff
@@ -12,6 +12,7 @@
     %meta{ name: "language-server-url", content: Exercism.config.language_server_url }
     - if user_signed_in?
       %meta{ name: "bug-reports-url", content: Exercism::Routes.api_bug_reports_url(auth_token: current_user.auth_tokens.first.to_s) }
```
Comment (iHiD) [inline app/views/layouts/application.html.haml]: This `auth_token` thing shouldn't be needed now.

File: app/views/layouts/application.html.haml [hunk 6]
```diff
@@ -12,6 +12,7 @@
...
+      %meta{ name: "parse-markdown-url", content: Exercism::Routes.api_parse_markdown_url }
```
Comment (iHiD) [inline app/views/layouts/application.html.haml]: I don't think any of these need to rely on `user_signed_in`. Let's remove that check, and just always render them in head. Makes it easier to consistently cache <head> when I add caching.

**Review by iHiD — APPROVED** [review]
Approved pending my comments :)

File: app/views/test/components/common/markdown_editor/show.html.haml (hunk 3 above)
Comment (kntsoriano) [reply app/views/test/components/common/markdown_editor/show.html.haml]: I think we should configure the id per context, so it reads better when viewing the keys on localstorage. I just generated a random id for here as this is just a test page.

**Comment by kntsoriano** [comment]
@ErikSchierboom Thanks. I've added the integration test and also added a `catch` block to show an error message when the API returns an error.

**Comment by ErikSchierboom** [comment]
@kntsoriano Great!

**Pull request merged**